Moved and revised expressions reference (#19744)
dotnet/docs: docs/csharp/language-reference/operators/index.md

---
title: "C# operators and expressions - C# reference"

> description: "Learn about C# operators and expressions, operator precedence and operator associativity"

Reviewer: nit

```suggestion
description: "Learn about C# operators and expressions, operator precedence, and operator associativity"
```

  - "expressions [C#]"
ms.assetid: 0301e31f-22ad-49af-ac3c-d5eae7f0ac43
---
# C# operators and expressions (C# reference)

> C# provides a number of operators. Many of them are supported by the [built-in types](../builtin-types/built-in-types.md) and allow you to perform basic operations with values of those types. For example, [arithmetic operators](arithmetic-operators.md) perform arithmetic operations with numeric operands and [Boolean logical operators](boolean-logical-operators.md) perform logical operations with [`bool`](../builtin-types/bool.md) operands. Others can be used in a broader scope. For example, you can use [conditional operator `?:`](conditional-operator.md) or [`switch` expression](switch-expression.md) to select which operand expression to evaluate based on a Boolean condition or a matched pattern, respectively.

Reviewer 2: The last two sentences seem unnecessary and are hard to follow without examples.  I would consider omitting them.

On the other hand, I think the mention of operator overloading is important here and shouldn't be dropped.

> C# provides a number of operators. Many of them are supported by the [built-in types](../builtin-types/built-in-types.md) and allow you to perform basic operations with values of those types. For example, [arithmetic operators](arithmetic-operators.md) perform arithmetic operations with numeric operands and [Boolean logical operators](boolean-logical-operators.md) perform logical operations with [`bool`](../builtin-types/bool.md) operands. Others can be used in a broader scope. For example, you can use [conditional operator `?:`](conditional-operator.md) or [`switch` expression](switch-expression.md) to select which operand expression to evaluate based on a Boolean condition or a matched pattern, respectively.

Reviewer 3: I agree with @tdykstra. This is a hard overview to get right. I think you're running into the same concerns we had trying to organize the operators into groups. How about this:

```suggestion
C# provides a number of operators. Many of them are supported by the [built-in types](../builtin-types/built-in-types.md) and allow you to perform basic operations with values of those types. For example, [arithmetic operators](arithmetic-operators.md) perform arithmetic operations with numeric operands and [Boolean logical operators](boolean-logical-operators.md) perform logical operations with [`bool`](../builtin-types/bool.md) operands. Other operators do not form obvious groups. For example, you can use [conditional operator `?:`](conditional-operator.md) or [`switch` expression](switch-expression.md) to select which operand expression to evaluate based on a Boolean condition or a matched pattern, respectively. Certain operators can be [overloaded](operator-overloading.md). With operator overloading, you can specify the operator behavior for the operands of a user-defined type. 
```

> The simplest C# expressions are literals (which may be [integer](../builtin-types/integral-numeric-types.md#integer-literals) or [real](../builtin-types/floating-point-numeric-types.md#real-literals) numbers, [characters](../builtin-types/char.md#literals) or [strings](../builtin-types/reference-types.md#the-string-type), [Boolean](../builtin-types/bool.md) `true` or `false`) and simple names (which usually denote variables). With use of operators, you get more complex expressions. Operator [precedence](#operator-precedence) and [associativity](#operator-associativity) determine the order in which the operations in an expression are performed. You can use parentheses to change the order of evaluation imposed by operator precedence and associativity.

Reviewer 2: char instances aren't really "characters".
"simple names" also seems like inexact language and leaves the reader wondering what "simple names" are that don't denote variables.

It isn't necessary at this point to enumerate all of the kinds of literals or names -- perhaps simplify this intro text as something like "The simplest C# expressions are literals and variable names. These can be combined into complex expressions by using operators."

> The simplest C# expressions are literals (which may be [integer](../builtin-types/integral-numeric-types.md#integer-literals) or [real](../builtin-types/floating-point-numeric-types.md#real-literals) numbers, [characters](../builtin-types/char.md#literals) or [strings](../builtin-types/reference-types.md#the-string-types), [Boolean](../builtin-types/bool.md) `true` or `false`) and simple names (which usually denote variables). With use of operators, you get more complex expressions. Operator [precedence](#operator-precedence) and [associativity](#operator-associativity) determine the order in which the operations in an expression are performed. You can use parentheses to change the order of evaluation imposed by operator precedence and associativity.

Reviewer: ```suggestion
The simplest C# expressions are literals (which may be [integer](../builtin-types/integral-numeric-types.md#integer-literals) or [real](../builtin-types/floating-point-numeric-types.md#real-literals) numbers, [characters](../builtin-types/char.md#literals) or [strings](../builtin-types/reference-types.md#the-string-type), [Boolean](../builtin-types/bool.md) `true` or `false`) and simple names (which usually denote variables). With use of operators, you get more complex expressions. Operator [precedence](#operator-precedence) and [associativity](#operator-associativity) determine the order in which the operations in an expression are performed. You can use parentheses to change the order of evaluation imposed by operator precedence and associativity.
```

In the following code, the right-hand operands of assignments are examples of expressions:

[!code-csharp[expression examples](snippets/Overview.cs#Expressions)]

> Additionally, C# provides the following forms of expressions:

Reviewer 2: Or make it sound less like a comprehensive list -- "Here are some other kinds of expressions that C# provides:"

Reviewer 3: I agree. Let's avoid making this a comprehensive list.

- [Interpolated string expressions](../tokens/interpolated.md) that provide convenient syntax to create formatted strings
- [Lambda expressions](../../programming-guide/statements-expressions-operators/lambda-expressions.md) that allow you to create anonymous functions
- [Query expressions](../keywords/query-keywords.md) that allow you to use query capabilities directly in C#

> Typically, an expression produces a result. A [`void`](../builtin-types/void.md) method invocation is an example of an expression that doesn't produce a result. Any [method invocation](member-access-operators.md#invocation-expression-) is also an example of a *statement expression*, which is an expression that you can use as a statement, that is, in the following form:

Reviewer 2: "...in the following form:" -- after the preceding examples, the difference here seems to be the absence of an = sign, but "any method invocation" would also include those that have a return value.  Perhaps "is also an example of" should be "may be used as"?

Reviewer 3: This is one of the more confusing areas of the language spec (because `void` really isn't a type). Here's how I would rework this paragraph:

1. Typically, an expression produces a result.
1. A *statement expression* can be used as a statement, meaning the result (if any), is discarded.
1. Invoking a method whose return type is `void` must be used as a statement, because it does not produce a result.
1. The only expressions that can be used as statements are:
  - method invocation
  - `await`
  - object creation
  - assignment
  - pre and post increment and decrement

I think reworking these two paragraph along those lines will be more clear.

```csharp
statement-expression;
```

> Other statement expressions are [assignment](assignment-operator.md), [await](await.md), [new object instantiation](new-operator.md), [increment](arithmetic-operators.md#increment-operator-), and [decrement](arithmetic-operators.md#decrement-operator---) expressions. When a statement expression is used as a statement, its result, if any, is discarded.

Reviewer 2: The items identified as "statement expressions" link to articles that call them operators, which is likely to be confusing.  It might be worth making this list into a bulleted list with an expression example for each.

Reviewer 3: I think reworking these two paragraph per my previous comment may make examples unnecessary. What do you think @tdykstra ?

Reviewer 2: Some parts of your outline after "The only expressions that can be used as statements are:" don't seem obvious to me without examples to explain them:

  - method invocation: this seems obvious
  - `await`: I don't think of "await" as an expression but more like just a keyword you put in front of a method invocation.
  - object creation: I usually think of this as `var x = new Poco();` -- if you discard the result from `new Poco()` what have you accomplished.
  - assignment: same issue here -- assignment I think of as `a = b`, `b` being the expression and `a = b` being the statement; not clear why the later should be called an expression statement
  - pre and post increment and decrement - this one seems obvious

Author: @tdykstra @BillWagner thank you for your thoughts and suggestions. However, as this topic looks to be more confusing than I thought, I prefer to remove it at all instead of going into more explanations (which would make the article harder to follow). All in all, expression statements are discussed in the [Statements](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/statements) article. That appears to be a more proper place for them.

- [Bitwise and shift operators](bitwise-and-shift-operators.md) that perform bitwise or shift operations with operands of the integral types
- [Equality operators](equality-operators.md) that check if their operands are equal or not

Typically, you can [overload](operator-overloading.md) those operators, that is, specify the operator behavior for the operands of a user-defined type.

The simplest C# expressions are literals (for example, [integer](../builtin-types/integral-numeric-types.md#integer-literals) and [real](../builtin-types/floating-point-numeric-types.md#real-literals) numbers) and names of variables. You can combine them into complex expressions by using operators. Operator [precedence](#operator-precedence) and [associativity](#operator-associativity) determine the order in which the operations in an expression are performed. You can use parentheses to change the order of evaluation imposed by operator precedence and associativity.

In the following code, examples of expressions are at the right-hand side of assignments:

[!code-csharp[expression examples](snippets/Overview.cs#Expressions)]

> Typically, an expression produces a result and can be included into another expression. A [`void`](../builtin-types/void.md) method call is an example of an expression that doesn't produce a result. It can be used only as a [statement](../../programming-guide/statements-expressions-operators/statements.md), as the following example shows:

Reviewer 2: ```suggestion
Typically, an expression produces a result and can be included in another expression. A [`void`](../builtin-types/void.md) method call is an example of an expression that doesn't produce a result. It can be used only as a [statement](../../programming-guide/statements-expressions-operators/statements.md), as the following example shows:
```

```csharp
Console.WriteLine("Hello, world!");
```

Here are some other kinds of expressions that C# provides:

- [Interpolated string expressions](../tokens/interpolated.md) that provide convenient syntax to create formatted strings:

>   [!code-csharp-interactive[interpolated string](snippets/Overview.cs#InterpolatedString)]

Reviewer 2: If you used the new syntax here, I believe you could highlight the line(s) referred to in the text, i.e., in this case the `var message =` line.

Reviewer: If highlighting by line number will be added, an *access-by-line.txt* file will be needed.

Author: @tdykstra thanks for a suggestion, but I'll leave it as it is now: the snippet is really small.
